File: leetcode/graph/694_Number_of_Distinct_Islands/codes/dfs.cpp
class Solution {
public:
    int numDistinctIslands(vector<vector<int>>& grid) {
        set<vector<pair<int,int>>> shapes;
        int m = grid.size(), n = grid[0].size();

        function<void(int, int, int, int, vector<pair<int,int>>&)> dfs = [&](int r0, int c0, int r, int c, vector<pair<int,int>>& shape) {
            if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == 0) return;
            grid[r][c] = 0;
            shape.emplace_back(r - r0, c - c0);
            dfs(r0, c0, r+1, c, shape);
            dfs(r0, c0, r-1, c, shape);
            dfs(r0, c0, r, c+1, shape);
            dfs(r0, c0, r, c-1, shape);
        };

        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                if (grid[i][j] == 1) {
                    vector<pair<int,int>> shape;
                    dfs(i, j, i, j, shape);
                    shapes.insert(shape);
                }

        return shapes.size();
    }
};

File: leetcode/graph/694_Number_of_Distinct_Islands/codes/dfs.py
from typing import List

class Solution:
    def numDistinctIslands(self, grid: List[List[int]]) -> int:
        def dfs(r, c, direction):
            if 0 <= r < len(grid) and 0 <= c < len(grid[0]) and grid[r][c] == 1:
                grid[r][c] = 0
                path.append(direction)
                dfs(r + 1, c, "D")
                dfs(r - 1, c, "U")
                dfs(r, c + 1, "R")  
                dfs(r, c - 1, "L") 
                path.append("B") 
        
        shapes = set()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    path = []
                    dfs(i, j, "S") 
                    shapes.add(tuple(path))
        return len(shapes)
